constantine/math/arithmetic/assembly/limbs_asm_mul_mont_arm64.nim

```diff
@@ -0,0 +1,444 @@
+    #   for j=0 to N-1
+    # 		(A,t[j])  := t[j] + a[j]*b[i] + A
+    #
+    # for 4 limbs, implicit column-wise carries
+    #
+    # t[0]     = t[0] + (a[0]*b[i]).lo
+    # t[1]     = t[1] + (a[1]*b[i]).lo + (a[0]*b[i]).hi
+    # t[2]     = t[2] + (a[2]*b[i]).lo + (a[1]*b[i]).hi
+    # t[3]     = t[3] + (a[3]*b[i]).lo + (a[2]*b[i]).hi
+    # overflow =                         (a[3]*b[i]).hi
+    #
+    # or
+    #
+    # t[0]     = t[0] + (a[0]*b[i]).lo
+    # t[1]     = t[1] + (a[0]*b[i]).hi + (a[1]*b[i]).lo
+    # t[2]     = t[2] + (a[2]*b[i]).lo + (a[1]*b[i]).hi
+    # t[3]     = t[3] + (a[2]*b[i]).hi + (a[3]*b[i]).lo
+    # overflow =    carry              + (a[3]*b[i]).hi
+    #
+    # Depending if we chain lo/hi or even/odd
+    # The even/odd carry chain is more likely to be optimized via μops-fusion
+    # as it's common to compute the full product. That said:
+    # - it's annoying if the number of limbs is odd with edge conditions.
+    # - fusion is not listed https://dougallj.github.io/applecpu/firestorm.html
+    #   "Other tested patterns are not fused, including adrp + add, mov + movk, mul + umulh, and udiv + msub."
+
+    ctx.mov A, xzr
+    ctx.ldr bi, b[i]
+    if i == 0:
+      for j in 0 ..< N:
+        ctx.mul t[j], aa[j], bi
+    else:
+      ctx.mulloadd_co(t[0], aa[0], bi, t[0])
+      for j in 1 ..< N:
+        ctx.mulloadd_cio(t[j], aa[j], bi, t[j])
+      ctx.adc A, xzr, xzr                        # assumes N > 1
+
+    ctx.mulhiadd_co(t[1], aa[0], bi, t[1])       # assumes N > 1
+    for j in 2 ..< N:
+      ctx.mulhiadd_cio(t[j], aa[j-1], bi, t[j])
+    ctx.mulhiadd_ci(A, aa[N-1], bi, A)
+
+    # Reduction
+    # -------------------------------
+    #   m := t[0]*m0ninv mod W
+    #
+    # 	C,_ := t[0] + m*M[0]
+    # 	for j=1 to N-1
+    # 		(C,t[j-1]) := t[j] + m*M[j] + C
+    #   t[N-1] = C + A
+    #
+    # for 4 limbs, implicit column-wise carries
+    #    _  = t[0] + (m*M[0]).lo
+    #  t[0] = t[1] + (m*M[1]).lo + (m*M[0]).hi
+    #  t[1] = t[2] + (m*M[2]).lo + (m*M[1]).hi
+    #  t[2] = t[3] + (m*M[3]).lo + (m*M[2]).hi
+    #  t[3] = A + carry          + (m*M[3]).hi
+    #
+    # or
+    #
+    #    _  = t[0] + (m*M[0]).lo
+    #  t[0] = t[1] + (m*M[0]).hi + (m*M[1]).lo
+    #  t[1] = t[2] + (m*M[2]).lo + (m*M[1]).hi
+    #  t[2] = t[3] + (m*M[2]).hi + (m*M[3]).lo
+    #  t[3] = A + carry          + (m*M[3]).hi
+
+    ctx.mul m, t[0], m0ninv
+    ctx.mul u, m, M[0]
+    ctx.cmn t[0], u         # TODO: bad latency chain, hopefully done parallel to prev loop
+    swap(u, v)
+
+    for j in 1 ..< N:
+      ctx.mulloadd_cio(t[j-1], m, M[j], t[j])
+    ctx.adc t[N-1], A, xzr
+
+    # assumes N > 1
+    ctx.mulhiadd_co(t[0], m, M[0], t[0])
+    for j in 1 ..< N-1:
+      ctx.mulhiadd_cio(t[j], m, M[j], t[j])
+    ctx.mulhiadd_ci(t[N-1], m, M[N-1], t[N-1])
+
+  if lazyReduce:
+    for i in 0 ..< N:
+      ctx.str t[i], r[i]
+  else:
+    # Final substraction
+    # we reuse the aa buffer
```

Comment:
Final subtraction to refactor in dedicated proc common with field addition and Montgomery reduction